### ISSUE TITLE
Upgrade HTTP to HTTPS instead of returning 426

### ISSUE THREAD

**Issue by sisp** [Mashape/kong#1946]
What's other people's opinion on adding the option to redirect HTTP connections to HTTPS instead of simply returning ``426`` with a message to use HTTPS?

**Comment by Tieske** [comment]
redirecting is slow and makes users lazy (they are getting away with doing nothing). Imo it is better to be strict and force usage of https by failing hard.

Only as a temporary solution while migrating from http to https a redirect makes sense to me.

**Comment by samrose** [comment]
I agree with the principle to be strict and force usage. But in case you cannot, here is what I did to resolve this in my case:

I created a template very similar to what is discussed in https://getkong.org/docs/0.9.x/configuration/#custom-nginx-configuration

I added the following to it which worked in my case:

```   
    server {
        listen 80 default_server;
        listen [::]:80 default_server;
        server_name _;
        return 301 https://$host$request_uri;
    }
```

Then just as discussed at https://getkong.org/docs/0.9.x/configuration/#custom-nginx-configuration 

you run your kong instance with `kong start --nginx-conf custom_nginx.template`

Kong will then use your server block rules in nginx

**Comment by Tieske** [comment]
@samrose thx for sharing!

**Comment by fabiorogeriosj** [comment]
this work for me, thx @samrose :)

**Comment by p0pr0ck5** [comment]
Closing, thanks @samrose for the snippet!

**Comment by thibaultcha** [comment]
Just as a side note to all of you involved in this discussion, if you are indeed using a custom Nginx template for this reason (or another), I strongly suggest that you check out #2355, which makes configuring Nginx easier, and invite you to your thoughts on it. We're actively looking for feedback. Thanks!

**Comment by andy9775** [comment]
Thanks for the snippet  @samrose - works great. I was also wondering if there is a way to perform a redirect from www to non-www while also doing an https upgrade? I'm fairly new to Nginx so I'm having trouble doing the redirect.

Basically, I want to upgrade HTTP to HTTPS and redirect www.example.com to example.com so:
http://www.example.com -> https://example.com
https://www.example.com -> https://example.com
http://example.com -> https://example.com

**Comment by samrose** [comment]
@andy9775 see this:

https://stackoverflow.com/a/7958540/996156


Best practice for nginx is to define a separate server for www the first example there would work if you just include 'https'

**Comment by andy9775** [comment]
@samrose doesn't seem to be working for me. Here's my template file

```
worker_processes ${{NGINX_WORKER_PROCESSES}}; # can be set by kong.conf
daemon ${{NGINX_DAEMON}};                     # can be set by kong.conf

pid pids/nginx.pid;                      # this setting is mandatory
error_log logs/error.log ${{LOG_LEVEL}}; # can be set by kong.conf

events {
    use epoll; 
    multi_accept on;
}

http {
    # include default Kong Nginx config
    include 'nginx-kong.conf';

    # upgrade all requests to https
    server {
        listen 8000 default_server;
        listen [::]:8000 default_server;
        server_name _;
        return 301 https://$host$request_uri;
    }

    server {
        listen 8000;
        server_name www.example.com;
        return 301 https://example.com$request_uri;
    }
}
```

I'm also using Docker.

**Comment by samrose** [comment]
@andy9775 yes, your requests are always going to get picked up by the first server block with that config. You will need to change it to something more like what is in the https://stackoverflow.com/a/7958540/996156 and drop the generic redirect from my original example. I am low on time to be able to layout a working example, but there is one at the SO link

**Comment by andy9775** [comment]
@samrose no worries. I think I get what you're saying now. Thanks!

**Comment by nbalsamo** [comment]
Shot in the dark, but has anyone been able to accomplish this on Heroku? For whatever reason, something simple like @samrose config change does not work.

**Comment by vinceve** [comment]
> redirecting is slow and makes users lazy (they are getting away with doing nothing). Imo it is better to be strict and force usage of https by failing hard.

I agree this is good if you use Kong only for internal stuff. It gives you quick feedback and you see immediately what's wrong. I don't agree when it comes to end users. Most people don't know even what https is. So they are going to see a message. Please use https.

We're Kong power users and we have some internal stuff configured in it, but also some public websites. And we love it. But we're facing an issue that we can't force SSL.

We can't use the nginx redirect to HTTPS template as some of our internal stuff doesn't work over https, and it also forces the server certificate, not the ones in Kong. The template is an all or nothing solution. I guess we need some more granular option, to redirect some public websites.

**Comment by PepperTeasdale** [comment]
I'm in a similar situation to @vinceve , though would definitely prefer to not change our applications as a workaround. One thought was to create a plugin that redirects to HTTPS, but can be installed on routes and services. Would anyone know if this is the kind of thing you can do with a custom plugin? Is there a repo of community-supported custom plugins somewhere? I've found a few random one-offs, but imagine this isn't the most uncommon use case. Would love direction on any of these things. Thanks! (btw if I end up creating this plugin, will be open-sourcing).

**Comment by coopr** [comment]
You'll find a bunch of community-published Kong plugins here https://luarocks.org/search?q=kong (Note that I'm not sure if you'll find plugins that address your particular use case)

**Comment by PepperTeasdale** [comment]
Making a plugin to do this on a service-by-service basis turned out to be fairly trivial. I'm hoping my company can open-source our solution, but until then this repo would be a good jumping off point for someone trying to do something similar: https://github.com/jicong/kong-aliyun-http-filter (thanks for @coopr , found this in that lua rocks search you linked)

**Comment by gaetanfl** [comment]
Another option is to use the `pre-function` plugin with the following function attached to a route or a service
```lua
local scheme = kong.request.get_scheme()
if scheme == "http" then
  local host = kong.request.get_host()
  local query = kong.request.get_path_with_query()
  return kong.response.exit(302,"https://" .. host ..query)
end
```

**Comment by hbagdi** [comment]
To share an update here, with https://github.com/Kong/kong/pull/4424 merged in, Kong 1.2 comes with support for sending HTTPS redirect out of the box in Kong.
